provider/azure.go

```diff
@@ -0,0 +1,456 @@
+	p.updateAzureRecords(updated)
+	return nil
+}
+
+func (p *AzureProvider) azureZones() ([]dns.Zone, error) {
+	log.Debug("Retrieving Azure DNS zones.")
+
+	var zones []dns.Zone
+	list, err := p.zonesClient.ListByResourceGroup(p.resourceGroup, nil)
+	if err != nil {
+		return zones, err
+	}
+
+	for list.Value != nil && len(*list.Value) > 0 {
+		for _, zone := range *list.Value {
+			if zone.Name != nil && strings.HasSuffix(*zone.Name, p.domainFilter) {
+				zones = append(zones, zone)
+			}
+		}
+
+		list, err = p.zonesClient.ListNextResults(list)
+		if err != nil {
+			return zones, err
+		}
+	}
+	log.Debugf("Found %d Azure DNS zone(s).", len(zones))
+	return zones, nil
+}
+
+func (p *AzureProvider) iterateRecords(zoneName string, callback func(dns.RecordSet) bool) error {
+	log.Debugf("Retrieving Azure DNS records for zone '%s'.", zoneName)
+
+	list, err := p.recordsClient.ListByDNSZone(p.resourceGroup, zoneName, nil)
+	if err != nil {
+		return err
+	}
+
+	done := false
+	for list.Value != nil && len(*list.Value) > 0 {
+		for _, recordSet := range *list.Value {
+			if !callback(recordSet) {
+				done = true
+				break
+			}
+		}
+
+		if done {
+			break
+		}
+
+		list, err = p.recordsClient.ListByDNSZoneNextResults(list)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func formatDNSName(recordName, zoneName string) string {
+	if recordName == "@" {
+		return zoneName
+	}
+	return fmt.Sprintf("%s.%s", recordName, zoneName)
+}
+
+func extractTarget(recordType string, recordSet *dns.RecordSet) string {
+	properties := recordSet.RecordSetProperties
+	if properties == nil {
+		return ""
+	}
+
+	switch dns.RecordType(recordType) {
+	case dns.A:
+		aRecords := properties.ARecords
+		if aRecords != nil && len(*aRecords) > 0 && (*aRecords)[0].Ipv4Address != nil {
+			return *(*aRecords)[0].Ipv4Address
+		}
+	case dns.CNAME:
+		cnameRecord := properties.CnameRecord
+		if cnameRecord != nil && cnameRecord.Cname != nil {
+			return *cnameRecord.Cname
+		}
+	case dns.TXT:
+		txtRecords := properties.TxtRecords
+		if txtRecords != nil && len(*txtRecords) > 0 && (*txtRecords)[0].Value != nil {
+			values := (*txtRecords)[0].Value
+			if values != nil && len(*values) > 0 {
+				return (*values)[0]
+			}
+		}
+	}
+	return ""
+}
+
+type azureChangeMap map[*dns.Zone][]*endpoint.Endpoint
+
+func mapAzureChanges(zones []dns.Zone, changes *plan.Changes) (azureChangeMap, azureChangeMap) {
+	ignored := map[string]bool{}
+	deleted := azureChangeMap{}
+	updated := azureChangeMap{}
+
+	mapChange := func(changeMap azureChangeMap, change *endpoint.Endpoint) {
+		zone := findAzureZone(zones, change.DNSName)
+		if zone == nil {
+			if _, ok := ignored[change.DNSName]; !ok {
+				ignored[change.DNSName] = true
+				log.Infof("Ignoring changes to '%s' because a suitable Azure DNS zone was not found.", change.DNSName)
+			}
+			return
+		}
+		// Change unspecified record types to address
+		if change.RecordType == "" {
+			change.RecordType = "A"
```

Comment:
shouldn't this use `suitableType` function from provider/provider.go ? RecordType should be decided on target, i.e. IP -> A, Hostname -> CNAME etc